add ProtoV6ProviderFactories to TestCase

kmoe opened the pull request:
In order to support protocol v6 providers in the test suite, we add another
field to `TestCase` along the same lines as `ProtoV5ProviderFactories`. This is
plumbed through as yet another argument in `runProviderCommand`. We hope that by
the time another protocol major version is released, a new, more flexible test
framework will be available.

`plugin.ServeOpts` and `plugin.DebugServe` also expected all provider servers to be
v5 servers, so we also add the `GRPCProviderV6Func` field to `ServeOpts` to keep
it backwards compatible. If both `GRPCProviderFunc` and `GRPCProviderV6Func` are
set, the v6 server will be ignored.

Alternatively, we could panic if both are set.

fixes #737 

needs https://github.com/hashicorp/terraform-exec/pull/182

Tested with terraform-provider-corner at https://github.com/hashicorp/terraform-provider-corner/pull/18.

paddycarver commented on helper/resource/plugin.go:
```diff
@@ -12,13 +12,14 @@ import (
 )
 
-func runProviderCommand(t testing.T, f func() error, wd *plugintest.WorkingDir, factories map[string]func() (*schema.Provider, error), v5factories map[string]func() (tfprotov5.ProviderServer, error)) error {
+func runProviderCommand(t testing.T, f func() error, wd *plugintest.WorkingDir, factories map[string]func() (*schema.Provider, error), v5factories map[string]func() (tfprotov5.ProviderServer, error), v6factories map[string]func() (tfprotov6.ProviderServer, error)) error {
```
At some point (not right now) we probably want to refactor this to take a "factories" struct as an argument and just populate it with all these variations.

paddycarver reviewed (commented):
This all seems straightforward and like what I expected so far!

I think, ideally, we'd be able to test the version of Terraform we were using and throw a more helpful error if the factories in use weren't supported by that version of Terraform, and it would probably be helpful to have a built-in SkipFunc that skips tests based on Terraform versions, but neither of those are directly related to this PR, so we can tackle that work separately.

kmoe commented:
Added https://github.com/hashicorp/terraform-plugin-sdk/issues/776 for the above suggestion, which I agree with. For now I have added a note to the docs for `ProtoV6ProviderFactories`. Also refactored along the lines discussed.

The tests in https://github.com/hashicorp/terraform-provider-corner/pull/18 are passing and that PR will be updated once this one is merged and released.

paddycarver reviewed (approved):
Solid, yeah, let's do it!
